cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/io/cdap/cdap/datapipeline/oauth/OAuthStore.java

```diff
@@ -0,0 +1,222 @@
+  private final TransactionRunner transactionRunner;
+  private final SecureStore secureStore;
+  private final SecureStoreManager secureStoreManager;
+
+  public static final StructuredTableId TABLE_ID = new StructuredTableId("oauth");
+  public static final StructuredTableSpecification TABLE_SPEC = new StructuredTableSpecification.Builder()
+      .withId(TABLE_ID)
+      .withFields(Fields.stringType(OAUTH_PROVIDER_COL),
+                  Fields.stringType(LOGIN_URL_COL),
+                  Fields.stringType(TOKEN_REFRESH_URL_COL))
+      .withPrimaryKeys(OAUTH_PROVIDER_COL)
+      .build();
+
+  public OAuthStore(
+      TransactionRunner transactionRunner,
+      SecureStore secureStore,
+      SecureStoreManager secureStoreManager) {
+    this.transactionRunner = transactionRunner;
+    this.secureStore = secureStore;
+    this.secureStoreManager = secureStoreManager;
+  }
+
+  /**
+   * Create/update an OAuth provider.
+   *
+   * @param oauthProvider {@link OAuthProvider} to write
+   * @throws OAuthStoreException if the write fails
+   */
+  public void writeProvider(OAuthProvider oauthProvider) throws OAuthStoreException {
+    try {
+      TransactionRunners.run(transactionRunner, context -> {
+        StructuredTable table = context.getTable(TABLE_ID);
+        table.upsert(getRow(oauthProvider));
+
+      }, TableNotFoundException.class, InvalidFieldException.class);
+    } catch (TableNotFoundException e) {
+      throw new OAuthStoreException("OAuth provider table not found", e);
+    } catch (InvalidFieldException e) {
+      throw new OAuthStoreException("OAuth provider object fields do not match table", e);
+    }
+
+    String namespace = NamespaceId.SYSTEM.getNamespace();
+    try {
+      secureStoreManager.put(
+          namespace,
+          getClientCredsKey(oauthProvider.getName()),
+          GSON.toJson(oauthProvider.getClientCredentials()),
+          "OAuth client creds",
+          EMPTY_MAP);
+    } catch (IOException e) {
+      throw new OAuthStoreException("Failed to write to OAuth provider secure storage", e);
+    } catch (Exception e) {
+      throw new OAuthStoreException("Namespace \"" + namespace + "\" does not exist", e);
+    }
+  }
+
+  /**
+   * Get an OAuth provider.
+   *
+   * @param name name of {@link OAuthProvider} to read
+   * @throws OAuthStoreException if the read fails
+   */
+  public Optional<OAuthProvider> getProvider(String name) throws OAuthStoreException {
+    OAuthClientCredentials clientCreds;
+    try {
+      String clientCredsJson = new String(
+          secureStore.get(NamespaceId.SYSTEM.getNamespace(), getClientCredsKey(name)).get(),
+          StandardCharsets.UTF_8);
+      clientCreds = GSON.fromJson(clientCredsJson, OAuthClientCredentials.class);
+    } catch (IOException e) {
+      throw new OAuthStoreException("Failed to read from OAuth provider secure storage", e);
+    } catch (Exception e) {
+      return Optional.empty();
```

Comment:
Why return an empty optional instead of throwing exception?

Reply:
I want to differentiate between lookup failure due to the entry not existing versus failure due to some error from reading from the store. Attempting to get the auth URL of a nonexistent provider returns a 404 not found, attempting to get the auth URL of a provider that might exist but there's some exception thrown while trying to retrieve it returns 500 internal error.

Reply:
The SecureStore.get() would actually throw an instance of `NotFoundException` if either the key or the namespace is not found, while throwing a general `Exception` for other cases. The documentation on the interface is not very clear about this point. It's better to do a:

```
try {
} catch (IOException e) {
   ...
} catch (NotFoundException e) {
  return Optional.empty();
} catch (Exception e) {
  throw new OAuthStoreException(..., e);
}
```